### PR TITLE
Do not use invasive lists to store CachePeers

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
Using invasive lists for CachePeer objects gives no advantages, but
requires maintaining non-standard error-prone code that leads to
excessive locking and associated memory overheads.

Also fixed a neighbors_init() bug that resulted in accessing an already
deleted "looks like this host" CachePeer object.

### CODE REVIEW

File: src/CachePeer.h [hunk 1]
```diff
@@ -227,6 +228,45 @@ class CachePeer
     void countFailure();
 };
 
+/// cache_peer configuration storage
+class CachePeers
```
Comment (eduard-bagdasaryan) [inline src/CachePeer.h:232]: We can move this new class into a separate header/source if needed.

File: src/CachePeer.cc [hunk 2]
```diff
@@ -131,3 +136,421 @@ operator <<(std::ostream &os, const CachePeer &p)
     return os << p.name;
 }
 
+/**
+ * utility function to prevent getservbyname() being called with a numeric value
+ * on Windows at least it returns garage results.
+ */
+static bool
+isUnsignedNumeric(const char *str, size_t len)
```
Comment (eduard-bagdasaryan) [inline src/CachePeer.cc:144]: I copied all this code 'as-is' from cache_cf.cc, without any refactoring. Should we polish it?

File: src/CachePeer.h [hunk 3]
```diff
@@ -227,6 +228,45 @@ class CachePeer
...
+
+    /// deletes a CachePeer object
+    void remove(CachePeer *);
+
+    const_iterator firstPing() const;
+    void advanceFirstPing();
+
+private:
+    CachePeerList cachePeers; ///< the list of parsed CachePeer objects
+    // TODO: document this field
```
Comment (eduard-bagdasaryan) [inline src/CachePeer.h:264]: I could not document this field (and related methods) because the old first_ping variable lacked any documentation - it is unclear what its purpose was.

File: src/CachePeer.cc [hunk 4]
```diff
@@ -131,3 +136,421 @@ operator <<(std::ostream &os, const CachePeer &p)
...
+    auto it = cachePeers.begin();
+    std::advance(it, firstPing_);
+    return it;
+}
+
+void
+CachePeers::advanceFirstPing()
+{
+    assert(firstPing_ <= size());
+    if (++firstPing_== size())
```
Comment (rousskov) [inline src/CachePeer.cc:531]: Can we merge this method with firstPing() and rename the result to nextPollStart()? It would be simpler and even a bit safer code AFAICT (e.g., no need to reset and no danger to forget to reset firstPing_ when removing cache peers).

**Review by rousskov — CHANGES_REQUESTED** [review]
Partial review to help you make progress.

File: src/CachePeer.h (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/CachePeer.h]: Ok, moved.

File: src/CachePeer.cc [hunk 5]
```diff
@@ -131,3 +136,421 @@ operator <<(std::ostream &os, const CachePeer &p)
...
+    for (auto it = cachePeers.begin(); it != cachePeers.end(); ++it) {
+        if (it->get() == p) {
+            cachePeers.erase(it);
+            break;
+        }
+    }
+    firstPing_ = 0;
+}
+
+CachePeers &
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/CachePeer.cc]: A caller that removed CachePeer needed a non-constant reference. I adjusted it, creating a dedicated function and then made this function return const reference.

Also, since first I made our begin()/end() methods non-const (as standard ones), this did not work with for-range loops (for-range loops do not use cbegin()/cend()). So to fix those loops I had to make begin()/end() const and removed (unused) cbegin()/cend().

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (96df01a).

File: src/CachePeers.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+private:
+    CachePeerList cachePeers; ///< the list of parsed CachePeer objects
+    /// approximate starting point for the next neighborsUdpPing() peer poll
+    size_t firstPing_ = 0;
+};
+
+const CachePeers &CurrentCachePeers();
+
+void NeighborRemove(CachePeer *);
```
Comment (eduard-bagdasaryan) [inline src/CachePeers.h:54]: We probably should rename this function into sth like `RemoveCachePeer()`, but I decided keep it for now (except the first letter register) to avoid confusion during reviews.

File: src/neighbors.cc [hunk 7]
```diff
@@ -616,9 +565,8 @@ neighborsUdpPing(HttpRequest * request,
 
     reqnum = icpSetCacheKey((const cache_key *)entry->key);
 
-    for (i = 0, p = first_ping; i++ < Config.npeers; p = p->next) {
-        if (p == nullptr)
-            p = Config.peers;
+    for (auto peer = CurrentCachePeers().nextPollStart(); peer != CurrentCachePeers().end(); ++peer) {
+        auto p = peer->get();
```
Comment (rousskov) [inline src/neighbors.cc:569]: AFAICT, this new loop will often miss cache peers. We want to iterate _all_ peers, and not just `[firstPing, end)` peers. Think round-robin from an arbitrary starting point...

The easiest way to implement this will probably result in something like this:

```suggestion
    for (size_t i = 0; i < Config.peers->size(); ++i) {
        const auto p = Config.peers->nextPeerToPoll().get();
```

... and keep the original Config.peers nil check above intact. This approach also avoids ugly const-casting in nextPeerToPoll(). The method will Assure() that at least one peer exists and will always return a peer.

----

With a lot more work, we could make it nicer, but I do _not_ think this improvement is worth our time spent on it:

```suggestion
    for (const auto peer:  Config.peers->peersToPoll()) {
        const auto p = peer.get();
```

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/neighbors.cc (hunk 7 above)
Comment (eduard-bagdasaryan) [reply src/neighbors.cc]: You are right, I mistakenly thought that the loop should poll only from first_ping (discarding the initial peers). I could not figure out how to adjust this with a single nextPeerToPoll() method - so in the solution I posted there are still two methods: nextPeerToPoll() and nextPollStart(). I have not tested it yet.

File: src/CachePeers.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#include "mem/PoolingAllocator.h"
+
+#include <memory>
+#include <vector>
+
+/// cache_peer configuration storage
+class CachePeers
+{
+public:
+    using CachePeerList = std::vector< std::unique_ptr<CachePeer>, PoolingAllocator< std::unique_ptr<CachePeer> > >;
```
Comment (rousskov) [inline src/CachePeers.h:22]: To avoid a false implication that this is a list (or any other specific type):

```suggestion
    using Storage = std::vector< std::unique_ptr<CachePeer>, PoolingAllocator< std::unique_ptr<CachePeer> > >;
```

File: src/CachePeers.h [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    /// the CachePeer used first in neighborsUdpPing() peer poll
+    const_iterator nextPollStart();
+
+    /// a CachePeer used next in neighborsUdpPing() peer poll
+    const_iterator nextPeerToPoll();
+
+private:
+    void setNextPing(size_t &) const;
+
+    CachePeerList cachePeers; ///< the list of parsed CachePeer objects
```
Comment (rousskov) [inline src/CachePeers.h:52]: To match the other suggestion:

```suggestion
    Storage storage; ///< cache_peers in configuration/parsing order
```

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/CachePeer.cc [hunk 10]
```diff
@@ -131,3 +136,421 @@ operator <<(std::ostream &os, const CachePeer &p)
...
+            storeAppendPrintf(entry, "neighbor_type_domain %s %s %s\n",
+                              p->host,
+                              peer_type_str(t->type),
+                              t->domain);
+        }
+    }
+}
+
+void
+CachePeers::parse(ConfigParser &)
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/CachePeer.cc]: Understood. I suspect it was a mistake to move this parsing and dumping code in this PR. The move hides the various small changes you had to make to integrate that old code with the new CachePeers class _and_ pressures us into polishing this ugly official code as well as converting it to use Configuration::Component API. I am sorry for not foreseeing these problems and asking you to move this code to CachePeers.cc.

Please move that code back to src/cache_cf.cc where it came from. CachePeers.cc will just host the new code or the code we have to move.

**Review by rousskov — CHANGES_REQUESTED** [review]

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (3f32f2d).

File: src/CachePeers.cc [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+    Assure(size());
+    if (currentPing_ == firstPing_) {
+        setNextPing(firstPing_);
+        currentPing_ = firstPing_;
+        return end();
+    }
+    auto it = cachePeers.begin() + currentPing_;
+    setNextPing(currentPing_);
+    return it;
+}
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/CachePeers.cc:21]: Done. I liked your idea with `peersPinged_ % size()` - the code becomes quite simple with just a single counter.

File: src/CachePeers.cc [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    const auto pos = peersPinged_ % size();
+    ++peersPinged_;
+    return storage.begin() + pos;
+}
+
+void
+CachePeers::remove(CachePeer *p)
+{
+    for (auto it = storage.begin(); it != storage.end(); ++it) {
+        if (it->get() == p) {
```
Comment (rousskov) [inline src/CachePeers.cc:27]: If you have not already, please check that we cannot use, in the order of preference:

* std::find(..., p) + storage.erase(pos) instead of this explicit loop (which also hides an important "not found" clause better)
* `*it == p`

File: src/CachePeers.cc [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        return peers;
+    }
+    return *Config.cachePeers;
+}
+
+void
+NeighborRemove(CachePeer *peer)
+{
+    Assure(Config.cachePeers);
+    Config.cachePeers->remove(peer);
```
Comment (rousskov) [inline src/CachePeers.cc:48]: This assertion seems inconsistent with ignoring a failure to find the peer in remove(). The official code does not guarantee removal AFAICT. If we want to preserve that behavior in this PR:

```suggestion
    if (Config.cachePeers)
        Config.cachePeers->removeIfFound(peer);
```

If we want to change that behavior, then this assertion is correct, but we need to change the remove() method to assert if the peer is not found.

File: src/neighbors.cc [hunk 14]
```diff
@@ -571,7 +524,7 @@ neighbors_init(void)
                 debugs(15, DBG_IMPORTANT, "WARNING: Peer looks like this host." <<
                        Debug::Extra << "Ignoring cache_peer " << *thisPeer);
 
-                neighborRemove(thisPeer);
+                NeighborRemove(thisPeer.get());
```
Comment (rousskov) [inline src/neighbors.cc:527]: Does removing items from the being-iterated vector constitute undefined behavior?

Comment (rousskov) [inline src/neighbors.cc:527]: Does the official code crash or trigger UB if the inner/HttpPortList loop continues after a neighborRemove() call? The current PR code has that problem, but I want to understand whether we are introducing a new bug here or preserving an old one.

This change request is _unrelated_ to the "removing items from the being-iterated vector" question in another change request.

File: src/neighbors.cc [hunk 15]
```diff
@@ -594,16 +545,14 @@ neighborsUdpPing(HttpRequest * request,
...
-    int i;
     int reqnum = 0;
     int flags;
     int peers_pinged = 0;
     int parent_timeout = 0, parent_exprep = 0;
     int sibling_timeout = 0, sibling_exprep = 0;
     int mcast_timeout = 0, mcast_exprep = 0;
 
-    if (Config.peers == nullptr)
+    if (!CurrentCachePeers().size())
```
Comment (rousskov) [inline src/neighbors.cc:555]: Let's undo this change if it is not necessary:

```suggestion
    if (Config.peers == nullptr)
```

File: src/peer_select.cc [hunk 16]
```diff
@@ -887,7 +888,7 @@ PeerSelector::selectAllParents()
      * simply are not configured to handle the request.
      */
     /* Add default parent as a last resort */
-    if ((p = getDefaultParent(this))) {
+    if (auto p = getDefaultParent(this)) {
```
Comment (rousskov) [inline src/peer_select.cc:891]:
```suggestion
    if (const auto p = getDefaultParent(this)) {
```

File: src/peer_sourcehash.cc [hunk 17]
```diff
@@ -125,7 +125,7 @@ peerSourceHashInit(void)
 
     for (k = 1; k <= K; ++k) {
         double Kk1 = (double) (K - k + 1);
-        p = sourcehash_peers[k - 1];
+        auto p = sourcehash_peers[k - 1];
```
Comment (rousskov) [inline src/peer_sourcehash.cc:128]:
```suggestion
        const auto p = sourcehash_peers[k - 1];
```

Please check for other candidates as well.

**Review by rousskov — CHANGES_REQUESTED** [review]

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (up to d1bfac6).

File: src/snmp_core.cc [hunk 18]
```diff
@@ -743,18 +743,14 @@ peer_Inst(oid * name, snint * len, mib_tree_entry * current, oid_ParseFn ** Fn)
...
         *len += 1;
     } else {
-        int no = name[current->len] ;
-        int i;
-        // Note: This works because the Config.peers keeps its index according to its position.
-        for ( i=0 ; peers && (i < no) ; peers = peers->next, ++i ) ;
-
-        if (peers) {
-            debugs(49, 6, "snmp peer_Inst: Encode peer #" << i);
+        const auto no = name[current->len];
```
Comment (rousskov) [inline src/snmp_core.cc:746]: Please do keep the old `int`-based declaration, the important (but wrong) source code comment, and the (refactored) silly loop above. This change makes `no` a `char`. That is worse than `int` because we do not know whether `char` is signed. There was a recent security vulnerability because of a similar conversion that went wrong. Also, we increment `no` below. Also, this official code is probably wrong because it should use `CachePeer::index` instead. Let's not bring all those concerns into this PR scope.

File: src/stat.cc [hunk 19]
```diff
@@ -12,6 +12,9 @@
 #include "AccessLogEntry.h"
 #include "CacheDigest.h"
 #include "CachePeer.h"
+#if USE_CACHE_DIGESTS
+#include "CachePeers.h"
+#endif
```
Comment (rousskov) [inline src/stat.cc:17]: Your call, but I would not imply that we should start creating complex conditions for our own header inclusion (where those conditions are not necessary to avoid some serious problems).

```suggestion
#include "CachePeers.h"
```

**Review by rousskov — CHANGES_REQUESTED** [review]
I think we are getting close to the final version, thank you.

File: src/CachePeers.cc [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    const auto pos = peersPinged_ % size();
+    ++peersPinged_;
+    return storage.begin() + pos;
+}
+
+void
+CachePeers::remove(CachePeer *p)
+{
+    storage.erase(std::remove_if(storage.begin(), storage.end(), [&](const auto &el) {
+                return el.get() == p; }), storage.end());
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/CachePeers.cc:27]: > the current PR code appears to remove up to two elements, the one equal to p (using std::remove_if()) and the next one (using std::vector::erase()

No, `std::remove_if()` does not remove elements physically, in just moves elements [around](https://en.cppreference.com/w/cpp/algorithm/remove) within a container, returning an iterator to the beginning of the first element to be deleted (should be passed to erase()). It is a standard erase-remove [idiom](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom).

[the remark this answers is not in the document]
Comment (rousskov) [reply src/CachePeers.cc:27]: Ah, right! Sorry. Well, I continue to hope that this code will not make it into the PR due to the other change request. :-)

File: src/neighbors.cc [hunk 21]
```diff
@@ -802,23 +753,17 @@ neighborsDigestSelect(PeerSelector *ps)
...
-        if (!p)
-            p = Config.peers;
+    for (size_t i = 0; i < Config.peers->size(); ++i) {
+        const auto p = Config.peers->nextPeerToPing()->get();
 
-        if (i == 1)
-            first_ping = p;
+        lookup_t lookup;
 
         lookup = peerDigestLookup(p, ps);
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/neighbors.cc]: Merged.

File: src/neighbors.cc [hunk 22]
```diff
@@ -1551,15 +1491,16 @@ dump_peer_options(StoreEntry * sentry, CachePeer * p)
 }
 
 static void
-dump_peers(StoreEntry * sentry, CachePeer * peers)
+dump_peers(StoreEntry *sentry)
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/neighbors.cc]: Ok, lets minimize these changes.

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (63bac48).

File: src/cache_cf.cc [hunk 23]
```diff
@@ -2068,12 +2069,16 @@ peer_type_str(const peer_t type)
...
+dump_peer(StoreEntry * entry, const char *name, const CachePeers *peers)
 {
+    if (!peers)
+        return;
+
     NeighborTypeDomainList *t;
     LOCAL_ARRAY(char, xname, 128);
 
-    while (p != nullptr) {
+    for (const auto &peer: (*peers)) {
```
Comment (rousskov) [inline src/cache_cf.cc:2080]: Jut to simplify:

```suggestion
    for (const auto &peer: *peers) {
```

File: src/cache_cf.cc [hunk 24]
```diff
@@ -2397,22 +2400,21 @@ parse_peer(CachePeer ** head)
...
 
-    *head = p;
+    p->index = (*peers)->size();
 
     peerClearRRStart();
 }
 
 static void
-free_peer(CachePeer ** P)
+free_peer(CachePeers **peers)
```
Comment (rousskov) [inline src/cache_cf.cc:2414]:
```suggestion
free_peer(CachePeers ** const peers)
```

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/CachePeers.h [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+class CachePeers
+{
+public:
+    using Storage = std::vector< std::unique_ptr<CachePeer>, PoolingAllocator< std::unique_ptr<CachePeer> > >;
+    using const_iterator = Storage::const_iterator;
+
+    /// appends a CachePeer object to the storage
+    void add(CachePeer *p) { storage.emplace_back(p); }
+    /// deletes a CachePeer object
+    void remove(const CachePeer *);
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/CachePeers.h]: Ok.

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (dcae79a2).

**Comment by rousskov** [comment]
This became official PR 1424.
